### ISSUE TITLE
Cheapest hours between x and y

### ISSUE THREAD

**Issue by heijjan** [T3m3z/spotprices2ha#11]
Hi,

I’ve been using this for roughly 6 months now and everything works like a charm. The last thing that I’m missing is the possibility to find x pieces of cheapest hours between a set range of time. This is mainly for charging my car and heating up the water. 

The ideal solution would be having an input field for required amount of hours, range start and range end for the hours so I could for example set it to charge my car between 22-07 during using the x pieces of cheapest hours. 

Creating the additional sensors / helpers is not the problem, but to find the code for the lookup. Any idea how to get this done?

**Comment by heijjan** [comment]
...so basically I'm asking about how it would be possible to reduce the amount of hours the cheapest period start helper is using for checking the cheapest hours:

![image](https://user-images.githubusercontent.com/111536037/226810196-8840fb83-e143-4647-bc8a-8228ae0ddcc9.png)

It would be great if there were cheapest periods start and cheapest period end attributes to control the lookup for the cheapest period.

**Comment by T3m3z** [comment]
I will consider adding this. I'm most probably going to refactor the code in the coming weeks (I hope that I have time for that) so maybe I could add this feature when I'm doing it.

**Comment by oskari** [comment]
> With some external help, I managed to update the sensor SHF Cheapest period start helper so that it takes two number helpers into account as the start and the end of the cheapest period check.

Wouldn't mind if you'd edit your screenshot to be an codeblock instead, makes it so that other people can you use your enhancement before (and if) it makes it to the official repo

**Comment by T3m3z** [comment]
I decided to rewrite a lot of the code. This allowed me to implement a new script that can be used to create complex automations if necessary. 


So please check this branch for the newer version (will be merged to master after I get experiences from the users):
https://github.com/T3m3z/spotprices2ha/tree/v0.2.0-draft

An example of an automation using this can be found from here:
https://github.com/T3m3z/spotprices2ha/blob/v0.2.0-draft/example-of-cheapest-period-automation.yaml

If you want to automate something, then follow these instructions:
1. Create a new time helper (Settings->Devices&Services->Helpers->Create Helper->"Date and/or time" + select "Time"). 
2. Create automation which updates this helper (see example-of-cheapest-period-automation.yaml for an example)
3. Then use the newly created helper in automations.

One use case could be some washing machine automation (get cheapest adjacent 3 hours between now and now+6 hours).

**Comment by jumakki** [comment]
Thanks for that script! It made my automation scripts much simpler. Everything seems to work correctly for me after updating to 0.2.0.
One feature request for new script.
Could it also calculate average price for returned period? I believe result can have multiple values, so in addition to result['start'] it could have average price in result['avg_price'] or something.
I wanted to show that also and was able to calculate it based on how it is done in spot-price.yaml, but would be nice to get that at the same time as start time.

```
 - service: input_number.set_value
    data:
      value: >-
        {%- set start = result['start'] -%} 
        {%- set stop = (start | as_datetime + timedelta(hours=4)) | as_timestamp -%}
        {{ state_attr("sensor.shf_data", "data") | selectattr("Timestamp", "ge", start) | selectattr("Timestamp", "lt", stop) | map(attribute="TotalPrice") | average | round(4) *100} }
    target:
      entity_id: input_number.heater_average_price

```

**Comment by jumakki** [comment]
There is something wrong with the price graph. Maybe related to it starting before midnight.
![Screenshot from 2023-09-18 08-25-17](https://github.com/T3m3z/spotprices2ha/assets/32912134/b19e3721-9481-4de3-9067-7e686e9219ee)

**Comment by jumakki** [comment]
It looks OK after 14:00 with next day prices included, but fails again in the morning.

**Comment by T3m3z** [comment]
I will try to check this today.

Br, Temez

ti 19. syysk. 2023 klo 8.49 jumakki ***@***.***> kirjoitti:

**Comment by T3m3z** [comment]
@jumakki : could you please try this new version https://raw.githubusercontent.com/T3m3z/spotprices2ha/v0.2.0-draft/apexchart-card-visualisations/current_electricity_price.yaml

**Comment by jumakki** [comment]
> @jumakki : could you please try this new version https://raw.githubusercontent.com/T3m3z/spotprices2ha/v0.2.0-draft/apexchart-card-visualisations/current_electricity_price.yaml

Looks good now, thanks!

**Comment by T3m3z** [comment]
Fixed by #21 .
